app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java

```diff
@@ -1107,4 +1116,47 @@ public boolean isDetailedNationalIDBuiltAndReady(int nationalID){
         return detailedPokemonShortList!=null && detailedPokemonShortList[nationalID]!=null;
     }
 
+    /**
+     * Constructs a sparse matrix of the type effectivenesses.
+     *
+     * This should be done on initiation of the {@link PokemonFactory}
+     *
+     * @return sparse matrix of the types effectiveness against each other
+     */
+    private double[][] buildTypeEffectivenessTable() {
+        int maxTypeID = getMaxTypeID();
+        double[][] typeEffectiveness = new double[maxTypeID][maxTypeID];
+
+        for (int i = 0; i < maxTypeID; i++) {
+            for (int j = 0; j < maxTypeID; j++) {
+                String[] selectionArg = {String.valueOf(i + 1), String.valueOf(j + 1)};
+                Cursor cursor = database.query(TYPE_EFFECTIVENESS_TABLE, null, FROM_TYPE_ID + "=?"
```

Comment:
This is O(n^2) time due to double for-loops above. Consider restructuring this query so that you load the entire table into memory and then format it into the Sparse Array.